document/src/vespa/document/base/idstring.h

```diff
@@ -13,108 +12,60 @@ namespace document {
-    static IdString::UP createIdString(const char *id, size_t sz);
     static LocationType makeLocation(vespalib::stringref s);
 
-    ~IdString();
-    IdString* clone() const override = 0;
+    explicit IdString(vespalib::stringref ns);
+    IdString();
 
-    virtual Type        getType() const = 0;
     vespalib::stringref getNamespace() const { return getComponent(0); }
-    virtual vespalib::stringref getNamespaceSpecific() const = 0;
-    virtual LocationType getLocation() const = 0;
-    virtual std::pair<int16_t, int64_t> getGidBitsOverride() const { return std::pair<int16_t, int64_t>(0, 0); }
-    virtual bool hasDocType() const { return false; }
-    virtual vespalib::stringref getDocType() const { return ""; }
-    virtual bool hasNumber() const { return false; }
-    virtual uint64_t getNumber() const { return 0; }
-    virtual bool hasGroup() const { return false; }
-    virtual vespalib::stringref getGroup() const { return ""; }
+    bool hasDocType() const { return size(1 != 0); }
+    vespalib::stringref getDocType() const  { return getComponent(1); }
+    LocationType getLocation() const  { return _location; }
+    bool hasNumber() const  { return _has_number; }
+    uint64_t getNumber() const  { return _location; }
+    bool hasGroup() const  { return _groupOffset != 0; }
+    vespalib::stringref getGroup() const  {
+        return vespalib::stringref(getRawId().c_str() + _groupOffset, offset(3) - _groupOffset - 1);
+    }
+    vespalib::stringref getNamespaceSpecific() const { return getComponent(3); }
 
     bool operator==(const IdString& other) const
         { return toString() == other.toString(); }
 
-    const vespalib::string & toString() const;
+    const vespalib::string & toString() const { return _rawId; }
 
-protected:
-    IdString(uint32_t maxComponents, uint32_t namespaceOffset, vespalib::stringref rawId);
+private:
     size_t offset(size_t index) const { return _offsets[index]; }
-    size_t size(size_t index) const { return _offsets[index+1] - _offsets[index] - 1; }
+    size_t size(size_t index) const { return std::max(0, _offsets[index+1] - _offsets[index] - 1); }
     vespalib::stringref getComponent(size_t index) const { return vespalib::stringref(_rawId.c_str() + offset(index), size(index)); }
     const vespalib::string & getRawId() const { return _rawId; }
-    virtual void validate() const;
-    size_t getNumComponents() const { return _offsets.numComponents(); }
 
-private:
     class Offsets {
     public:
-        Offsets(uint32_t maxComponents, uint32_t first, vespalib::stringref id);
-        uint16_t first() const { return _offsets[0]; }
+        Offsets() = default;
+        uint16_t compute(vespalib::stringref id);
         uint16_t operator [] (size_t i) const { return _offsets[i]; }
-        size_t numComponents() const { return _numComponents; }
+        static const Offsets DefaultID;
     private:
+        Offsets(vespalib::stringref id);
         uint16_t _offsets[5];
```

Comment:
Consider exposing and using `MAX_COMPONENTS` here to reduce number of magical constants

Reply:
Done.